### PR TITLE
opentelemetry: create a new module

### PR DESCRIPTION
**Pull request by andriikushch**
**What**
This PR adds a new module to support OpenTelemetry tracing.

The new module wraps implementation provided by: [github.com/open-telemetry/opentelemetry-go](https://github.com/open-telemetry/opentelemetry-go) and made it possible to append it to the chain of handlers.

This module is compatible with specification: [opentelemetry-specification](https://github.com/open-telemetry/opentelemetry-specification).

**How to configure**

This module can be enabled in Caddyfile and configured with env variables mentioned in the specification. Please see more information in README.md in this PR.

**Known limitation**

- The current implementation supports only **grpc** otlp protocol.
- It supports only the following propagators **tracecontext,baggage**

Some other limitations are mentioned here: https://github.com/open-telemetry/opentelemetry-go/issues/1698

### CODE REVIEW

File: modules/caddyhttp/opentelemetry/README.md
Comment (hairyhenderson) [inline modules/caddyhttp/opentelemetry/README.md:19]: The metrics and logs parts of the OTel spec are non-stable, but Tracing is stable so I'm not sure if this statement is accurate. Either way, it adds a bit of confusion that I'm not sure is necessary in this context.

File: modules/caddyhttp/opentelemetry/README.md
Comment (hairyhenderson) [inline modules/caddyhttp/opentelemetry/README.md:15]: Can we link directly to the environment variables spec?

```suggestion
It can be configured using the environment variables defined
by the [OpenTelemetry Environment Variable Specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md).
```

Comment (andriikushch) [reply modules/caddyhttp/opentelemetry/README.md:15]: Perfectly makes sense to me. Good point! 👍

File: modules/caddyhttp/opentelemetry/README.md
Comment (hairyhenderson) [inline modules/caddyhttp/opentelemetry/README.md:22]: How about:

```suggestion
These environment variables are required by this module:
- `OTEL_EXPORTER_OTLP_PROTOCOL` or `OTEL_EXPORTER_OTLP_TRACES_PROTOCOL`
- `OTEL_PROPAGATORS`
```

(I'm not sure exactly what you mean though - maybe adjust this)

File: modules/caddyhttp/opentelemetry/tracer.go
Comment (hairyhenderson) [inline modules/caddyhttp/opentelemetry/tracer.go:45]:
```suggestion
	ErrUnsupportedTracesProtocol = errors.New("unsupported opentelemetry traces protocol")
```

File: modules/caddyhttp/opentelemetry/README.md
Comment (hairyhenderson) [inline modules/caddyhttp/opentelemetry/README.md:49]:
```suggestion
handle /myHandler {
```

Comment (andriikushch) [reply modules/caddyhttp/opentelemetry/README.md:49]: Good catch 👍

**Comment by hairyhenderson** [comment]
Hi @andriikushch, thank you so much for working on this! I don't have a lot of time to put any more time into reviewing this today unfortunately, and I'll be on vacation with limited Internet access for the rest of the week.

I think it'd be useful to see some of the attributes added by https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp added to this as well (though you probably won't be able to use that directly).

Also, I'm not sure how much I like the fact that certain environment variables are required. Most OTel-enabled applications I've dealt with will default to using OTLP with gRPC, so I think defaulting to that is totally reasonable.

Since many users of Caddy are not yet familiar with OpenTelemetry, or with Distributed Tracing in general, I think it would be very helpful to write a short tutorial article describing how to get started, similar to https://caddyserver.com/docs/metrics.

I had a few other vague thoughts but they escape me now, and I need to get going - again, thank you for working on this!

File: modules/caddyhttp/opentelemetry/README.md
Comment (andriikushch) [reply modules/caddyhttp/opentelemetry/README.md:22]: You understood it absolutely correctly 😃 . And I think the proposed by you change sounds clearer. 👍

File: modules/caddyhttp/opentelemetry/README.md
Comment (andriikushch) [reply modules/caddyhttp/opentelemetry/README.md:19]: The reason why I put here this statement is this ticket: https://github.com/open-telemetry/opentelemetry-go/issues/1698. 

I just wanted to emphasize that customers should not expect that, for example, variable OTEL_TRACES_SAMPLER will change the behavior of the tracing.

What do you think, does it make sense to add this clarification or it is better to remove this line?

File: modules/caddyhttp/opentelemetry/tracer.go
Comment (andriikushch) [reply modules/caddyhttp/opentelemetry/tracer.go:45]: 👍

File: modules/caddyhttp/opentelemetry/README.md
Comment (francislavoie) [inline modules/caddyhttp/opentelemetry/README.md:4]:
```suggestion
as a `caddyhttp.MiddlewareHandler` and can be chained into a list of other handlers.
```

**Comment by andriikushch** [comment]
JFYI: For the next few days, I will not have Internet access and I will be glad to implement/discuss all suggestions after the 9th of October.

**Comment by hairyhenderson** [comment]
@andriikushch sorry for the excessively long silence on this! I'm starting to review it now.

I have a few initial observations, though a more thorough review is under way:
- I see that you've put the module as a subdirectory of `modules/caddyhttp/` - was this intentional? Shouldn't it be directly in `modules/`?
- At least initially, I'd like to support _only_ OTLP over gRPC. I think if you just call [`otlptracegrpc.New(ctx)`](https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/otlp/otlptrace/otlptracegrpc/exporter.go#L24) you shouldn't need to do any environment variable handling - this should all be done for you by the client (see https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/otlp/otlptrace/README.md#configuration)

**Comment by francislavoie** [comment]
It's implemented as an HTTP middleware/handler, so I think it's ok in `caddyhttp`, unless it's planned to also be used in other contexts.

File: modules/caddyhttp/opentelemetry/module.go
Comment (hairyhenderson) [inline modules/caddyhttp/opentelemetry/module.go:17]: Naming-wise, I think it may be better to use a more generic term like `tracing`. We're not likely to support any other tracing APIs or SDKs. The OpenTelemetry project also supports more than just tracing, and I think it would be best to explicitly focus this module on tracing:

```suggestion
	caddy.RegisterModule(Tracing{})
	httpcaddyfile.RegisterHandlerDirective("tracing", parseCaddyfile)
```

File: modules/caddyhttp/opentelemetry/module.go
Comment (hairyhenderson) [inline modules/caddyhttp/opentelemetry/module.go:25]: Why not simply `span`?

```suggestion
	SpanName string `json:"span"`
```

File: modules/caddyhttp/opentelemetry/module.go
Comment (hairyhenderson) [inline modules/caddyhttp/opentelemetry/module.go:89]: I think `span` is probably better

```suggestion
		"span": &ot.SpanName,
```

File: modules/caddyhttp/opentelemetry/tracer.go
Comment (andriikushch) [reply modules/caddyhttp/opentelemetry/tracer.go:144]: I checked it also and you are right 👍 . I will change it.

File: modules/caddyhttp/opentelemetry/tracer.go
Comment (andriikushch) [reply modules/caddyhttp/opentelemetry/tracer.go:74]: Valid point! 👍  Will be changed in another commit.

File: modules/caddyhttp/opentelemetry/tracer.go
Comment (andriikushch) [reply modules/caddyhttp/opentelemetry/tracer.go:92]: 👍

File: modules/caddyhttp/opentelemetry/tracer.go
Comment (andriikushch) [reply modules/caddyhttp/opentelemetry/tracer.go:40]: Agree 👍

File: modules/caddyhttp/opentelemetry/tracer.go
Comment (andriikushch) [reply modules/caddyhttp/opentelemetry/tracer.go:153]: You are absolutely right, I will remove this function completely. 👍

File: modules/caddyhttp/opentelemetry/tracer.go
Comment (andriikushch) [reply modules/caddyhttp/opentelemetry/tracer.go]: 👍

File: modules/caddyhttp/opentelemetry/tracer.go
Comment (andriikushch) [reply modules/caddyhttp/opentelemetry/tracer.go:223]: I will do it 👍

File: modules/caddyhttp/opentelemetry/tracer.go
Comment (andriikushch) [reply modules/caddyhttp/opentelemetry/tracer.go:86]: 👍 Agree

**Comment by andriikushch** [comment]
Hi @hairyhenderson! 

Thank you very much for such a detailed review and comments! 👍 

Answering your question: @francislavoie is right, my intention was to implement HTTP middleware, this is a reason why I decided to put it here.

Regarding all other comments, I think I implemented all the suggestions and I like that now the code looks much cleaner and straightforward.

Also, I cleaned up tests and updated versions in `go.mod`.

Additionally, I made some manual tests for a simple scenario of tracing and all seems to work fine 😃  

Thank you once more for the great code review!

File: modules/caddyhttp/tracing/README.md
Comment (hairyhenderson) [inline modules/caddyhttp/tracing/README.md:6]:
```suggestion
When enabled, it will propagate an existing trace context or initialize a new one.
```

File: modules/caddyhttp/tracing/module.go
Comment (hairyhenderson) [inline modules/caddyhttp/tracing/module.go:22]: some grammar/working tweaks:

```suggestion
// Tracing implements an HTTP handler that adds support for distributed tracing,
// using OpenTelemetry. This module is responsible for the injection and
// propagation of the trace context. Configure this module via environment
// variables (see https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md).
// Some values can be overwritten in the configuration file.
```

File: modules/caddyhttp/tracing/module.go
Comment (hairyhenderson) [inline modules/caddyhttp/tracing/module.go:24]:
```suggestion
	// SpanName is a span name. It should follow the naming guidelines here:
	// https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#span
```

File: modules/caddyhttp/tracing/module.go
Comment (hairyhenderson) [inline modules/caddyhttp/tracing/module.go:51]:
```suggestion
// Cleanup implements caddy.CleanerUpper and closes any idle connections. It
// calls Shutdown method for a trace provider https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#shutdown.
```

**Comment by hairyhenderson** [comment]
Ok, @andriikushch I think this is pretty much ready to go. Thanks for your patience!

Some things that are still missing, which can probably be added in future PRs:
- I'd like to see a `traceID` field added to all log messages when there's a trace in the context
  - this is required for systems that can correlate logs with traces (for example in Grafana with Tempo & Loki we have a logs-to-traces feature that uses this field)
- The metrics middleware should be able to observe histograms with [exemplars](https://grafana.com/docs/grafana/latest/datasources/prometheus/#configuring-exemplars)
  - this is a simple change (using the [`ObserveWithExemplar`](https://pkg.go.dev/github.com/prometheus/client_golang@v1.11.0/prometheus#ExemplarObserver) method instead of `Observe` [like here](https://github.com/caddyserver/caddy/blob/e7457b43e4703080ae8713ada798ce3e20b83690/modules/caddyhttp/metrics.go#L129) and whenever else a histogram is observed)
- Documentation for the `tracing` Caddyfile directive (this needs to go in https://github.com/caddyserver/website/tree/master/src/docs/markdown/caddyfile/directives)
- A tutorial on how to use this, like https://github.com/caddyserver/website/blob/master/src/docs/markdown/metrics.md

These changes _might_ require some changes to ensure that the span is available in-context for these, so I'm wondering if there's some way we can mark this as "experimental" (cc @mholt)?

But otherwise I think this is very close to merging - it'll provide great value even if it's only able to propagate traces!

**Comment by andriikushch** [comment]
Hi @hairyhenderson !

Thank you very much for your feedback. I agree that improvements might be added with the following PRs. I accepted the proposed changes and solved the conflict, so the branch is ready to merge.

Best regards,
Andrii

**Comment by hairyhenderson** [comment]
ok, this LGTM - @mholt do you have any questions about this?

@andriikushch looks this needs to be rebased again - I attempted to do it myself but I think you have the `Allow edits by maintainers` checkbox unchecked? (or maybe I don't have the right permissions 😂)

**Comment by hairyhenderson** [comment]
(I know @mholt has been quite swamped this week so this may take some time, but I would like his feedback given this is a fairly significant new feature)

**Comment by andriikushch** [comment]
Hi @hairyhenderson ,

For some reason, a checkbox `Allow edits by maintainers` is not visible, in this PR. 🤔 

Anyway, I updated a branch, so it can be merged again.

Please let me know if there is something I can help with merging this PR. 😃 

Best regards,
Andrii

**Comment by francislavoie** [comment]
@andriikushch I think it's because the PR branch comes from https://github.com/instana/caddy which isn't actually your fork, it's under another org. I think the PR would need to be from a branch on _your_ account for that setting to appear. Kinda annoying. It means we'll have to ask you to update your branch _just before_ we merge, cause otherwise we can't click the merge button I see this, basically:

image

**Comment by andriikushch** [comment]
Hi @francislavoie,

I think you are right 👍 . I have asked my colleagues to also check if there is any setting for this.

Meanwhile, I have updated a branch 😃

**Comment by mholt** [comment]
Thanks for keeping this updated. Still have one failing test, maybe that's just a fluke?

**Comment by francislavoie** [comment]
> Error: fatal: unable to access 'caddyserver/caddy': Could not resolve host: github.com

So yeah, just something broken on the test environment. Re-running.

File: modules/caddyhttp/tracing/README.md
Comment (mholt) [inline modules/caddyhttp/tracing/README.md:1]: Not against a readme here, but I think this is the only folder in this repo that has one (other than the root). Should we move this to the site instead so people will actually see it?

File: modules/caddyhttp/tracing/tracerprovider.go
Comment (mholt) [inline modules/caddyhttp/tracing/tracerprovider.go:50]: use `zap.Error()` instead of rendering the error string to the message if possible. `logger.Error("forcing flush", zap.Error(err))`

**Review by mholt — CHANGES_REQUESTED** [review]
Thanks for the work on this, @andriikushch -- looking good. Just a couple nits.

File: modules/caddyhttp/tracing/tracer.go
Comment (vibhavp) [inline modules/caddyhttp/tracing/tracer.go:83]: Would it be a good idea to use `r.WithContext(ctx)`, so that downstream plugins that use OpenTelemetry can add to this trace? In this manner, plugins would get to reuse the `TracerProvider` initialized by this plugin, instead of setting up their own redundant tracing pipeline.

**Comment by vibhavp** [comment]
Hello! I've created a PR on top of this branch that uses[ `otelhttp` ](https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp)to instrument HTTP requests, by passing the next handler and the associated error result in a "carrier" struct injected into the request's context. It's a slightly hack-y way of sending these values, but it lets us take advantage of the instrumentation code written and maintained by the opentelemetry-go community. I'm not sure if the code is completely idiomatic because of the way it [uses](https://github.com/instana/caddy/pull/1/files#diff-02ad5e853ecdb06eecf263e79a0880b0f51a383749bdd4171bca7ad4d84b12d8) `Context`, but any feedback on it would be great: https://github.com/instana/caddy/pull/1

File: modules/caddyhttp/tracing/tracerprovider.go
Comment (andriikushch) [reply modules/caddyhttp/tracing/tracerprovider.go]: Makes sense to me. Done.

File: modules/caddyhttp/tracing/README.md
Comment (andriikushch) [reply modules/caddyhttp/tracing/README.md:1]: @mholt, I agree with you. I remove it from here and will create a PR to the website repo.

**Comment by ostcar** [comment]
I am exited about this PR and hope, it will be merged (and released) soon. Thank you for your work!

It is possible with this PR to configure caddy for the [sampling](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#sampling)? We use caddy as proxy to our services. It would be nice, if caddy could be configured to trace only some request. For example [ratio based](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#traceidratiobased).

In the documentation of this PR I can't find anything about sampling. Do I have to use the environment variable `OTEL_TRACES_SAMPLER`? Would it be possible to add a configuration to the module, so the sampling can also be configured in the caddy-file? Or is this out of scope of this PR?

**Comment by andriikushch** [comment]
> Hello! I've created a PR on top of this branch that uses[ `otelhttp` ](https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp)to instrument HTTP requests, by passing the next handler and the associated error result in a "carrier" struct injected into the request's context. It's a slightly hack-y way of sending these values, but it lets us take advantage of the instrumentation code written and maintained by the opentelemetry-go community. I'm not sure if the code is completely idiomatic because of the way it [uses](https://github.com/instana/caddy/pull/1/files#diff-02ad5e853ecdb06eecf263e79a0880b0f51a383749bdd4171bca7ad4d84b12d8) `Context`, but any feedback on it would be great: [instana#1](https://github.com/instana/caddy/pull/1)

Hi, @vibhavp,

Thank you a lot for your contribution! I like the proposed improvement and already merged it in this PR.

Good job 👍!  Thank you!

**Comment by andriikushch** [comment]
> I am exited about this PR and hope, it will be merged (and released) soon. Thank you for your work!
> 
> It is possible with this PR to configure caddy for the [sampling](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#sampling)? We use caddy as proxy to our services. It would be nice, if caddy could be configured to trace only some request. For example [ratio based](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#traceidratiobased).
> 
> In the documentation of this PR I can't find anything about sampling. Do I have to use the environment variable `OTEL_TRACES_SAMPLER`? Would it be possible to add a configuration to the module, so the sampling can also be configured in the caddy-file? Or is this out of scope of this PR?

Hi, @ostcar, 

Thank you for your interest in the topic. You are absolutely right, this PR does not support the configuration of the sampling via Caddyfile. 

It was a design decision to use environment variables to configure the "opentelemetry" functionality. 

As far as I know, currently "opentelemetry" go library does not support the OTEL_TRACES_SAMPLER environment variable. Here is more info about why https://github.com/open-telemetry/opentelemetry-go/issues/1698.

Nevertheless, there is an open issue for this https://github.com/open-telemetry/opentelemetry-go/issues/2305 and already created PR https://github.com/open-telemetry/opentelemetry-go/pull/2517. It looks that soon this support will be added.

Best regards,
Andrii

File: modules/caddyhttp/tracing/module.go
Comment (mholt) [inline modules/caddyhttp/tracing/module.go:68]: Might be helpful to document the syntax here in the comment; I try to do this for most Caddyfile unmarshalers.

File: modules/caddyhttp/tracing/serve.go
Comment (mholt) [inline modules/caddyhttp/tracing/serve.go:1]: Does this file need to be in its own file? (It's fine, I suppose, but seems a bit random to put all by itself in a separate file.) I wonder if this whole thing should just go in tracer.go.

**Review by mholt — CHANGES_REQUESTED** [review]
Looking much better, thank you :tada: 

My remaining comments are minor nits, we can probably merge this shortly.

File: modules/caddyhttp/tracing/module.go
Comment (andriikushch) [reply modules/caddyhttp/tracing/module.go]: Hi @mholt! Thank you a lot for the quick feedback.

👍 I agree. I will add it, in a minute.

File: modules/caddyhttp/tracing/serve.go
Comment (andriikushch) [reply modules/caddyhttp/tracing/serve.go]: 👍 I agree there is no reason to have a separate file for this indeed. I will move it to "trace.go".  😃

**Comment by mholt** [comment]
Ok, great, this is looking good. Thanks very much for building this.

I pulled down the branch and built Caddy and this PR adds > 2 MB to the total size of the binary (about a 5% increase). I'm a bit sensitive to that now as we approach 50 MB 😬. That _seems_ like kind of a lot for tracing requests but I guess I'm not too surprised.

I was surprised, however, to see that this feature doesn't change _anything_ in existing Caddy files, let alone the core -- only adding files. (Yay for Caddy's extensible architecture!) I figured there'd have to be some code in the core of the HTTP module, like where the HTTP handler chain is compiled. (This isn't a bad thing, just an observation.)

Anyway, as I think on things, and as Caddy 2's standard distribution is finally beginning to stabilize, I wonder if this could be better had as a separate module first, and then we can bring it into the official/standard distribution if it's used by enough users. This will let us prove out the stability of dependencies (something I wish I had done before with other features) and reduce bloat and maintenance burden, while still serving everyone who wants to use tracing. (Especially since we already have logging and metrics in the standard distribution.)

We would still make it an official repo in the `caddyserver` org, and put you in charge of maintaining it. Then we can see how popular and necessary it is deemed by experience. It's much easier to bring something into the official repo than it is to deprecate/remove it if there are problems.

What do you think?

**Comment by cedricziel** [comment]
Hi @mholt , thanks for your extensive notes :) 

I am a product manager with Instana and Andrii asked me for some feedback here. Technically we (and probably other observability vendors as well) think observability needs to be built into the core of modern-day software to make them observable by default. We created this module in order to enable everyone running Caddy in production to integrate with their observability suite out of the box without a custom build.

Our preference here would definitely be an in-tree plugin shipped by default, so users of your helm-chart (https://github.com/caddyserver/ingress/issues/52) could enable it out-of-the box. However if you feel uncomfortable, we'd also be willing to go the external-plugin route to gain some traction.

**Comment by proffalken** [comment]
Just wanted to add my voice to this PR - I'm in full agreement with @cedricziel that observability should be a "core feature" rather than an "after thought" or "optional extra", however I'd rather see tracing as a plugin rather than not at all!

 The progress made here is fantastic, and if I can provide our customers with a recommendation for a lightweight, Open Source webserver that has both openmetrics and opentelemetry built in rather than having to filter out stuff from Nginx, Traefik, or Apache logs, then that would be amazing!

**Review by mholt — APPROVED** [review]
Ok, thanks for the feedback.

Now obviously this is a big change and I haven't scrutinized every line, nor have I personally tested it, but as it's purely additive and there seems to be strong consensus it should ship by default, this has my approval.

However, I think I'd like to mark this module as EXPERIMENTAL in the godoc comments so that in the documentation on our website it's clear to users that it could be changed or removed (like, brought out to a separate repository) at any time (for example, if there's major problems discovered). Just until we get confident/comfortable with things.

**Comment by proffalken** [comment]
Thank you @mholt, that's fantastic news!

**Comment by mholt** [comment]
If the merge conflicts are resolved before we tag 2.5 (in the next couple of days), we can include it in that release, otherwise it'll have to wait to 2.6.

**Comment by cedricziel** [comment]
Since Andrii is currently out, I took the liberty of resolving the conflicts :)

**Comment by mholt** [comment]
Awesome, thanks @cedricziel -- merging in just a moment.

**Pull request merged**
